lib/docker/executeDockerBuild.ts

```diff
@@ -256,47 +258,61 @@ async function dockerPush(images: string[],
 
-    let result = Success;
-
     if (await pushEnabled(gi, options)) {
 
-        // 1. run docker push
-        for (const image of images) {
-            result = await gi.spawn(
-                "docker",
-                ["push", image],
-                {
-                    env: {
-                        ...process.env,
-                        DOCKER_CONFIG: dockerConfigPath(options, gi.goalEvent),
+        // Login to registry(s)
+        await Promise.all(
+            options.registries.map(async r => {
+                const loginResult = await dockerLogin(r, options, gi);
+                if (loginResult.code !== 0) {
+                    throw new Error(`Failed login for registry => ${r.url}!`);
+                }
+            }),
+        );
+
+        await Promise.all(
+            images.map(async image => {
+                const pushResult = await gi.spawn(
+                    "docker",
+                    ["push", image],
+                    {
+                        env: {
+                            ...process.env,
+                            DOCKER_CONFIG: dockerConfigPath(options, gi.goalEvent),
+                        },
+                        log: gi.progressLog,
                     },
-                    log: gi.progressLog,
-                },
-            );
+                );
 
-            if (result && result.code !== 0) {
-                return result;
-            }
-        }
+                if (pushResult.code !== 0) {
+                    throw new Error(`Failed to push image ${image}!  Error (code: ${pushResult.code}) => ${pushResult.stderr}`);
+                }
+
+            }),
+        );
     } else {
         gi.progressLog.write("Skipping 'docker push'");
     }
 
-    return result;
+    return {code: 0};
 }
 
 export const DefaultDockerImageNameCreator: DockerImageNameCreator = async (p, sdmGoal, options, context) => {
     const name = cleanImageName(p.name);
-    const tags = [await readSdmVersion(sdmGoal.repo.owner, sdmGoal.repo.name,
-        sdmGoal.repo.providerId, sdmGoal.sha, sdmGoal.branch, context)];
+    const version =
+        await readSdmVersion(sdmGoal.repo.owner, sdmGoal.repo.name, sdmGoal.repo.providerId, sdmGoal.sha, sdmGoal.branch, context);
 
+    // If there are configured registries, set tags for each; otherwise return just version
+    const tags: string[] = [];
     const latestTag = await projectConfigurationValue<boolean>("docker.tag.latest", p, false);
-    if (latestTag && sdmGoal.branch === sdmGoal.push.repo.defaultBranch) {
-        tags.push("latest");
-    }
+    options.registries.map(r => {
+        tags.push(`${r.url}/${name}:${version}`);
+        if (latestTag && sdmGoal.branch === sdmGoal.push.repo.defaultBranch) {
+            tags.push(`${r.url}/${name}:latest`);
+        }
+    });
 
     return {
-        registry: options.registry,
         name,
         tags,
     };
```

Comment:
this should return an `Array<{ registry, name, tag }>`

Reply:
As discussed, will update